### DISCUSSION TITLE
fori_loop and jacrev: it works ! but I thought it wasn't

### DISCUSSION THREAD

**Discussion by jecampagne** [jax-ml/jax#14643] [Q&A] [answered]
Well,
Writing a demo nb and show that scan can perform jacrev but not fori_loop, I was astonished to discover that
```python
@jit
def g(x):

  def body(i,val):
    v1, v2, v3, a = val 
    return v1+a[i], v2+a[i]**2, v3+a[i]**3, a

  val_init=(0.,0.,0.,x)
  N=x.shape[0]
  res = jax.lax.fori_loop(0,N,body,val_init)
  return jnp.array(list(res[:-1]))
```
```
y = jax.random.uniform(jax.random.PRNGKey(10),shape=(100,))

g(y) # DeviceArray([51.85117211, 35.0536869 , 26.71050342], dtype=float64)
jnp.sum((jacfwd(g))(y), axis=1) # DeviceArray([100.        , 103.70234421, 105.1610607 ], dtype=float64)
jnp.sum((jacrev(g))(y), axis=1) # DeviceArray([100.        , 103.70234421, 105.1610607 ], dtype=float64)
```
Well when I read the doc, it is clear that when the fori_loop lower and upper index are known a priori (eg. static variables) then jax.lax.scan is used,
but here I was expecting that "N" is not a priori known?

Am I wrong?, or fori_loop has changed (I am running 0.3.25 on Colab) ?

Is there a simple use-case to demonstrated that fori_loop jacrev is not working but scan does? Thanks

**Comment by jakevdp** [comment, marked as answer]
Thanks for the question! N here is known at compile time, because array shapes are static. If you had used an array value in place of N, then it would be dynamic and your function would not be reverse-mode differentiable.

**Reply by jecampagne** [reply]
if the size of x change, remind me but JIT compilation is triggered, right?

**Reply by jecampagne** [reply]
Well; I am certainly bad but
```python
def h(x,N):
  def body(i,val):
    v1, v2, v3, a = val 
    return v1+a[i], v2+a[i]**2, v3+a[i]**3, a

  val_init=(0.,0.,0.,x)
  res = jax.lax.fori_loop(0,N,body,val_init)
  return jnp.array(list(res[:-1]))
```
```python
y = jax.random.uniform(jax.random.PRNGKey(10),shape=(100,))
jnp.sum((jacrev(h))(y, jnp.array([10])), axis=1)
```
is working?!

**Reply by jakevdp** [reply]
N is not traced in this function call, so it is static. If you were to JIT-compile the function h, then N would be traced and reverse mode autodiff would fail.

**Reply by jecampagne** [reply]
But if I use @jit fo h then it crashes at call
y = jax.random.uniform(jax.random.PRNGKey(10),shape=(100,))
h(y,jnp.array([10]))

UnfilteredStackTrace: TypeError: cond_fun must return a boolean scalar, but got output type(s) [ShapedArray(bool[1])].

although I see that it uses while... 

[/usr/local/lib/python3.8/dist-packages/jax/_src/lax/control_flow/loops.py](https://localhost:8080/#) in fori_loop(lower, upper, body_fun, init_val)
   1694   else:
-> 1695     _, _, result = while_loop(_fori_cond_fun, _fori_body_fun(body_fun),
   1696                               (lower, upper, init_val))

**Reply by jakevdp** [reply]
For example:
```python
import jax.numpy as jnp
import jax

@jax.jit
def h(x,N):
  def body(i,val):
    v1, v2, v3, a = val 
    return v1+a[i], v2+a[i]**2, v3+a[i]**3, a

  val_init=(0.,0.,0.,x)
  res = jax.lax.fori_loop(0,N,body,val_init)
  return jnp.array(list(res[:-1]))

y = jax.random.uniform(jax.random.PRNGKey(10),shape=(100,))
jax.jacrev(h)(y, 10)
```
```pytb
...
ValueError: Reverse-mode differentiation does not work for lax.while_loop or lax.fori_loop. Try using lax.scan instead.
```
This happens because under `jit`, `N` is traced and so `fori_loop` uses `while_loop` rather than `scan`, and `while_loop` is not reverse-mode differentiable.

**Reply by jecampagne** [reply]
ok I see. Thanks

**Reply by jecampagne** [reply]
Hi @jakevdp  Is there a lax.scan solution?  Well, I succeed but do not know if it is the best
```python
@partial(jit, static_argnums=(1,))
def h(x,N):
  
  tmp=x[:N]

  def body(a, i):
    return a, jnp.sum(a**i) 

  _, res = jax.lax.scan(body,tmp,jnp.array([1,2,3]))

  return res
```

BUT here we require @partial(jit, static_argnums=(1,)), which turns out to be a working solution for the fori_loop too even for the jacrev.

So I'm back to find an example that one do need to use scan.

**Reply by jecampagne** [reply]
Hi again @jakevdp , trying to find a fori_loop example where jacrev do not find solution, I was  digging into discussion (3850)[https://github.com/google/jax/discussions/3850] I though that I have found my example
which where reporting a check_grads pb. Although I do not see pbs and jacrev works for scan_fun
```python
@jit
def loop_fun(x, y):

    n = x.shape[0]

    def body(i, curr): 
        return curr + x[i] * y[i]

    return jax.lax.fori_loop(0, n, body, 0.) / n

@jit
def scan_fun(x, y):
    n = x.shape[0]

    def body(carry, x):
        curr, i = carry
        return (curr + x * y[i], i+1), None

    (res, _), _ = jax.lax.scan(body, (0., 0), x)
    return res / n
```
```
n = 50
m = 150
x = jnp.arange(n, dtype=jnp.float32)
y = jnp.arange(m, step=2., dtype=jnp.float32)

tmp1 = jax.jacrev(scan_fun)(x,y)
tmp2 = jax.jacrev(loop_fun)(x,y)

tmp1==tmp2
```

**Reply by jakevdp** [reply]
I’m not sure I quite understand what you’re asking, but keep in mind that if the `fori_loop` endpoints are static, it will be lowered to `scan`, and if they’re dynamic, it will be lowered to `while`. Your examples seem consistent with that.

**Reply by jecampagne** [reply]
Sorry to disturb you @jakevdp  I appreciate very much your comment, no doubt about that. I'm trying to get an example that show ones that he/she is forced to use `scan` instead of `fori_loop` when on needs  jacrev with JIT on, AND the partial(jit, argumet_static=...) is not possible.

**Reply by jakevdp** [reply]
I don't think you'll find any such example: the only time a `fori_loop` can be replaced by a `scan` is when the endpoints are static, and whenever the endpoints are static, `fori_loop` will lower to `scan`, and therefore will be compatible with jacrev.

**Reply by jecampagne** [reply]
So the slogan "always scan when you can!" is more a matter of efficiency more than impossibility to find a solution to get fori_loop working with JIT/jacrev ?

**Reply by jakevdp** [reply]
yes, I believe so

**Reply by jecampagne** [reply]
well may be the [discussions 3850](https://github.com/google/jax/discussions/3850) is obsolete.

**Reply by jakevdp** [reply]
Yes, I think it's probably obsolete – it was right around then that `fori_loop` was changed to automatically use scan when possible (#2414)

**Reply by jecampagne** [reply]
Ok thanks now I have a clearer view.
